File: go/AnomalyDetector/AnomalyDetectorQuickstart.go
package main

import (
	"bufio"
	"context"
	"encoding/csv"
	"fmt"
	"github.com/Azure/azure-sdk-for-go/services/preview/cognitiveservices/v1.0/anomalydetector"
	"github.com/Azure/go-autorest/autorest"
	"github.com/Azure/go-autorest/autorest/date"
	"io"
	"log"
	"os"
	"strconv"
	"time"
)

/*
This quickstart shows examples for:
	Detecting anomalies in a batch (in a whole series of points)
	Detecting anomalies in stream (latest data point)

Prerequisites:
	- Install libraries on the command line (or in IDE):
		go get github.com/Azure/azure-sdk-for-go/services/preview/cognitiveservices/v1.0/anomalydetector
		go get github.com/Azure/go-autorest/autorest
	- Set your Anomaly Dector subscription key and endpoint to your environment variables.
	- Add the request-data.csv to your local root folder

How to run (from command line):
	go run AnomalyDetectorQuickstart.go

References:
	Anomaly Detector documentation: https://docs.microsoft.com/en-us/azure/cognitive-services/anomaly-detector/
	SDK: https://godoc.org/github.com/Azure/azure-sdk-for-go/services/preview/cognitiveservices/v1.0/anomalydetector
	API: https://westus2.dev.cognitive.microsoft.com/docs/services/AnomalyDetector/operations/post-timeseries-entire-detect
*/

var key  = os.Getenv("ANOMALY_DETECTOR_SUBSCRIPTION_KEY")
var endpoint = os.Getenv("ANOMALY_DETECTOR_ENDPOINT")
// Spreadsheet data that will be converted to a series of points
var filePath = "./request-data.csv"

func main() {

	// Authenticate your client for all examples
	client := anomalydetector.New(endpoint)
	client.Authorizer = autorest.NewCognitiveServicesAuthorizer(key)

	// Get data from file , used for all examples
	var series []anomalydetector.Point
	csvFile, _ := os.Open(filePath)
	reader := csv.NewReader(bufio.NewReader(csvFile))
	for {
		line, err := reader.Read()
		if err == io.EOF { break } else if err != nil { log.Fatal(err) }

		timestamp, _ := time.Parse(time.RFC3339, line[0])
		value, _ := strconv.ParseFloat(line[1], 64)

		series = append(series, anomalydetector.Point{Timestamp: &date.Time{timestamp}, Value: &value})
	}

	// Request body for the API call in both batch and stream examples
	var request = anomalydetector.Request{Series: &series, Granularity: anomalydetector.Daily}

	/*
	ANOMALY DETECTOR - BATCH
	*/
	// Detect the anomalies
	fmt.Println()
	fmt.Println("Detecting anomalies in the entire series...")
	fmt.Println()

	// Make the call
	responseBatch, err := client.EntireDetect(context.Background(), request)
	if err != nil { log.Fatal("ERROR:", err) }

	var anomalies int
	for index, isAnomaly := range *responseBatch.IsAnomaly {
		if isAnomaly {
			fmt.Println("An anomaly is detected from the series at row:" + strconv.Itoa(index))
			anomalies++
		}
	}
	if (anomalies == 0) {
		fmt.Println("There are no anomalies detected from the series.")
	}
	/*
	END - ANOMALY DETECTOR - BATCH
	*/
	
	/*
	ANOMALY DETECTOR - STREAM
	*/
	fmt.Println()
	fmt.Println("Detecting the latest point of a series for anomalies...")
	fmt.Println()

	responseStream, err := client.LastDetect(context.Background(), request)
	if err != nil { log.Fatal("ERROR:", err) }

	if *responseStream.IsAnomaly {
		fmt.Println("The latest point, in row " + strconv.Itoa(len(series)) + ", is detected as an anomaly.")
	} else {
		fmt.Println("The latest point, in row " + strconv.Itoa(len(series)) + ", is not detected as an anomaly.")
	}
	/*
	END - ANOMALY DETECTOR - STREAM
	*/
}

File: javascript/AnomalyDetector/anomaly_detector_quickstart.js
// <imports>
'use strict'

const fs = require('fs');
const parse = require("csv-parse/lib/sync");
const { AnomalyDetectorClient } = require('@azure/ai-anomaly-detector');
const { AzureKeyCredential } = require('@azure/core-auth');
// </imports>

/**
 * This quickstart will detect anomolies in spreadsheet (.csv) data in two different ways:
 *  - Using the batch method (detects anomalies in entire series of data points)
 *  - Using the stream method (detects the latest data point in a series)
 * 
 * Prerequisites:
 *   - Install the following modules: 
 *       npm install csv-parse
 *       npm install @azure/core-auth
 *       npm install @azure/ai-anomaly-detector
 *   - Add your Anomaly Detector subscription key and endpoint to your environment variables
 *   - Add the request-data.csv file to your local root folder
 * 
 * Reference:
 *   Anomaly Detector documentation: https://docs.microsoft.com/en-us/azure/cognitive-services/anomaly-detector/
 *   SDK: https://docs.microsoft.com/en-us/javascript/api/overview/azure/cognitiveservices/anomalydetector?view=azure-node-latest
 *   API: https://westus2.dev.cognitive.microsoft.com/docs/services/AnomalyDetector/operations/post-timeseries-entire-detect
 */

//<vars>
// Spreadsheet with 2 columns and n rows.
let CSV_FILE = './request-data.csv';

// Authentication variables
let key = 'PASTE_YOUR_ANOMALY_DETECTOR_SUBSCRIPTION_KEY_HERE';
let endpoint = 'PASTE_YOUR_ANOMALY_DETECTOR_ENDPOINT_HERE';

// Points array for the request body
let points = [];
//</vars>

// <authentication>
let anomalyDetectorClient = new AnomalyDetectorClient(endpoint, new AzureKeyCredential(key));
// </authentication>

// <readFile>
function readFile() {
    let input = fs.readFileSync(CSV_FILE).toString();
    let parsed = parse(input, { skip_empty_lines: true });
    parsed.forEach(function (e) {
        points.push({ timestamp: new Date(e[0]), value: parseFloat(e[1]) });
    });
}
readFile()
// </readFile>

// <batchCall>
async function batchCall() {
    // Create request body for API call
    let body = { series: points, granularity: 'daily' }
    // Make the call to detect anomalies in whole series of points
    await anomalyDetectorClient.detectEntireSeries(body)
        .then((response) => {
            console.log("Batch (entire) anomaly detection):")
            for (let item = 0; item < response.isAnomaly.length; item++) {
                if (response.isAnomaly[item]) {
                    console.log("An anomaly was detected from the series, at row " + item)
                }
            }
        }).catch((error) => {
            console.log(error)
        })

}
batchCall()
// </batchCall>

// <lastDetection>
async function lastDetection() {

    let body = { series: points, granularity: 'daily' }
    // Make the call to detect anomalies in the latest point of a series
    await anomalyDetectorClient.detectLastPoint(body)
        .then((response) => {
            console.log("Latest point anomaly detection:")
            if (response.isAnomaly) {
                console.log("The latest point, in row " + points.length + ", is detected as an anomaly.")
            } else {
                console.log("The latest point, in row " + points.length + ", is not detected as an anomaly.")
            }
        }).catch((error) => {
            console.log(error)
        })
}
lastDetection()
// </lastDetection>

// <changePointDetection>
async function changePointDetection() {

    let body = { series: points, granularity: 'daily' }
    // get change point detect results
    await anomalyDetectorClient.detectChangePoint(body)
        .then((response) => {
            if (
                response.isChangePoint.some(function (changePoint) {
                    return changePoint === true;
                })
            ) {
                console.log("Change points were detected from the series at index:");
                response.isChangePoint.forEach(function (changePoint, index) {
                    if (changePoint === true) {
                        console.log(index);
                    }
                });
            } else {
                console.log("There is no change point detected from the series.");
            }
        }).catch((error) => {
            console.log(error)
        })
}
// </changePointDetection>
changePointDetection();